Reconfigure MetalK8s authn to support external backends(Keycloak)

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'dex', 'charts', 'oidc'

**Context**: 

See: [ticket](https://scality.atlassian.net/browse/RINGX-18)

**Summary**:

- Reconfigure Dex to add support for any external Identity Provider
- Add documentation on how to edit the CSC that handles authn config params

**Acceptance criteria**: 

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
charts: add support for an external IDP to Dex

This commit adds support for enabling an external IDP
provider like Keycloak using the Dex staticClients param.

This charts is re-rendered using
```
./charts/render.py dex --namespace metalk8s-auth \
 charts/dex.yaml charts/dex/ --service-config dex \
 metalk8s-dex-config metalk8s/addons/dex/config/dex.yaml \
 metalk8s-auth > salt/metalk8s/addons/dex/deployed/chart.sls
```

Ebaneck committed:
CSC: expose external OIDC client configuration

This commit exposes configuration params in the Dex CSC
for supporting external identity providers.

Here, Keycloak or a local deployment of the MetalK8s UI can be
plugged into Dex accordingly.

Ebaneck committed:
docs: add documentation on how to plug-in an external IDP

This commit adds documentation on how to add a new external Identity
Provider ontop of MetalK8s.

The new external IDP configuration is possible thanks to the exposure of Dex
`staticclient` configurations using CSC mechanism

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

TeddyAndrieux commented on charts/dex.yaml:
```diff
@@ -89,6 +90,11 @@ config:
     secret: "4lqK98NcsWG5qBRHJUqYM1"
+  - id: external-oidc-ui
+    name: '__var__(dex.spec.externalIDP.staticClient.name)'
+    redirectURIs:
+    - '__var__(dex.spec.externalIDP.staticClient.redirectURIs)'
```
To me `redirectURIs` is a list of URI so it's should NOT be in a list here 
```suggestion
    redirectURIs: '__var__(dex.spec.externalIDP.staticClient.redirectURIs)'
```

TeddyAndrieux replied:
Also here since it's a list we want/need to put a `| tojson` filter to this one

TeddyAndrieux commented on salt/metalk8s/addons/dex/deployed/chart.sls:
```diff
@@ -60,6 +61,10 @@ stringData:
       secret: 4lqK98NcsWG5qBRHJUqYM1
+    - id: external-oidc-ui
+      name: {% endraw -%}{{ dex.spec.externalIDP.staticClient.name }}{%- raw %}
+      redirectURIs:   {% endraw -%}{{ dex.spec.externalIDP.staticClient.redirectURIs | tojson }}{%- raw %}
```
Weird those triple space, but never mind it's not related to change of this PR

```suggestion
      redirectURIs: {% endraw -%}{{ dex.spec.externalIDP.staticClient.redirectURIs | tojson }}{%- raw %}
```

Ebaneck replied:
Likely from the render script.
Let me fix it

Ebaneck replied:
~~Going to backport the changes maybe later!~~

TeddyAndrieux commented on salt/metalk8s/addons/dex/deployed/chart.sls:
```diff
@@ -60,9 +61,13 @@ stringData:
       secret: 4lqK98NcsWG5qBRHJUqYM1
+    - id: external-oidc-ui
+      name: {% endraw -%}{{ dex.spec.externalIDP.staticClient.name }}{%- raw %}
+      redirectURIs: {% endraw -%}{{ dex.spec.externalIDP.staticClient.redirectURIs | tojson }}{%- raw %}
+      secret: ZXhhbXBsZS1hcHAtc2VjcmV0
     enablePasswordDB: {% endraw -%}{{ dex.spec.localuserstore.enabled }}{%- raw %}
     staticPasswords:
-      {% endraw -%}{{ dex.spec.localuserstore.userlist | tojson }}{%- raw %}
+    {% endraw -%}{{ dex.spec.localuserstore.userlist | tojson }}{%- raw %}
```
Not sure it will work like this right now

Ebaneck committed:
render: add support for substituting unquoted Jinja to YAML

This commit ensures when magically substituting unquoted Jinja to YAML
we respect spacing in and around the Jinja variable.

gdemonet commented:
Two major comments (we can't merge as-is):

- This is adding a **client** to Dex, not a **connector**: it's like allowing another IdP (such as Keycloak) to federate Dex users, which is the opposite of what we want
- Adding clients is a good feature (in case someone wants to add another control-plane UI, e.g. what we'd do for Solutions), but the implementation is wrong IMO: this PR hardcodes a single client, to which we can only add redirect URIs - we should instead allow to configure an arbitrary amount of clients to Dex, and let these clients configure their own secrets

On the topic of secrets, I don't think we should keep hardcoded secrets for our own clients anyway (they are open-sourced, hence vulnerable). Let's take a step back and rethink this PR offline.

gdemonet commented:
Let's close this in favor of #2781 if that's OK for you @Ebaneck

Ebaneck commented:
Closing due to https://github.com/scality/metalk8s/pull/2763#issuecomment-688694805
